ui/app/utils/clickhouse/datasets.server.ts

```diff
@@ -345,53 +176,79 @@ export async function countDatapointsForDatasetFunction(
   if (!function_type) {
     return null;
   }
-  const table =
-    function_type === "chat"
-      ? "ChatInferenceDatapoint"
-      : "JsonInferenceDatapoint";
-  const resultSet = await getClickhouseClient().query({
-    query: `SELECT toUInt32(count()) as count FROM {table:Identifier} WHERE dataset_name = {dataset_name:String} AND function_name = {function_name:String}`,
-    format: "JSONEachRow",
-    query_params: { dataset_name, function_name, table },
+  const dbClient = await getNativeDatabaseClient();
+  return dbClient.countDatapointsForDatasetFunction({
+    dataset_name,
+    function_name,
+    function_type,
   });
-  const rows = await resultSet.json<{ count: number }>();
-  const parsedRows = rows.map((row) => CountSchema.parse(row));
-  return parsedRows[0].count;
-}
-
-function validateDatasetName(dataset_name: string) {
-  if (dataset_name === "builder" || dataset_name.startsWith("tensorzero::")) {
-    throw new Error("Invalid dataset name");
-  }
 }
 
 export async function getAdjacentDatapointIds(
   dataset_name: string,
   datapoint_id: string,
-): Promise<AdjacentIds> {
-  const query = `
-    WITH DatasetIds AS (
-      SELECT toUInt128(id) as id_uint FROM ChatInferenceDatapoint WHERE dataset_name = {dataset_name:String}
-      UNION ALL
-      SELECT toUInt128(id) as id_uint FROM JsonInferenceDatapoint WHERE dataset_name = {dataset_name:String}
-    )
-    SELECT
-      NULLIF(
-      (SELECT uint_to_uuid(min(id_uint)) FROM DatasetIds WHERE id_uint > toUInt128({datapoint_id:UUID})),
-      toUUID('00000000-0000-0000-0000-000000000000')
-      ) as next_id,
-      NULLIF(
-        (SELECT uint_to_uuid(max(id_uint)) FROM DatasetIds WHERE id_uint < toUInt128({datapoint_id:UUID})),
-        toUUID('00000000-0000-0000-0000-000000000000')
-      ) as previous_id
-    FROM DatasetIds
-  `;
-  const resultSet = await getClickhouseClient().query({
-    query,
-    format: "JSONEachRow",
-    query_params: { dataset_name, datapoint_id },
+): Promise<AdjacentDatapointIds> {
+  const dbClient = await getNativeDatabaseClient();
+  return await dbClient.getAdjacentDatapointIds({
+    dataset_name,
+    datapoint_id,
   });
-  const rows = await resultSet.json<AdjacentIds>();
-  const parsedRows = rows.map((row) => adjacentIdsSchema.parse(row));
-  return parsedRows[0];
+}
+
+/**
+ * Converts a backend Datapoint to a frontend ParsedDatasetRow.
+ * This is used when receiving data from the backend API.
+ * TODO(shuyangli): Remove soon!
```

Comment:
we should ideally expose APIs to Node that return resolved data I think

Reply:
I do not think there is ever a reason to get a pre-resolved datapoint for the UI and not need the images and stuff.